### DISCUSSION TITLE
How do I use binary caching with different toolchain versions?

### DISCUSSION THREAD

**Discussion by nishanthkarthik** [microsoft/vcpkg#18563] [Q&A]
I am reading the binary caching spec at https://vcpkg.readthedocs.io/en/latest/specifications/binarycaching/ and the example at https://devblogs.microsoft.com/cppblog/vcpkg-accelerate-your-team-development-environment-with-binary-caching-and-manifests/

Consider this scenario

I am on ubuntu 18.04 with the default gcc-7 toolchain and I install clang on top of it. With clang as my compiler, I build a package and upload it to my nuget caching server. For the next build, I install gcc-8 on the same environment and clang will now pick up gcc-8 as the selected toolchain. The update to gcc-8 will also bump the system libstdc++ version. If I build the same package again, will vcpkg fetch the clang+gcc-7-toolchain build from the cache? 

From vcpkg's eyes, the architecture, OS and the compiler versions have not changed. The only thing that's different in the second build is the libstdc++ version.

tldr: Does binary caching also take care of the toolchain version providing libstdc++?

**Comment by PhoebeHui** [comment]
From https://github.com/microsoft/vcpkg/blob/master/docs/specifications/binarycaching.md#tracking-compilers, vcpkg do not track the compiler used, so I think the toolchain version changes doesn't affect the binary caching.

@dan-shaw, could you help double confirm?

**Comment by viromer** [comment]
These are the latest details about the [compiler tracking used in binary caching](https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md#implementation-notes-internal-details-subject-to-change-without-notice), internally vcpkg uses CMake to detect the compiler executables being used (see implementation in scripts/detect_compiler). I don't know if these will be enough to detect the changes in your environment (I'm personally inexperienced with Clang) but as the docs state, you can add extra information as comments in a triplet file, since the contents of the triplet file are hashed and used as part of the hash calculation for binary caching.

@ras0219-msft can you help with this question?

**Reply by knzivid** [reply]
bump?

**Reply by Osyotr** [reply]
The fact that vcpkg tracks compiler hash is a real bummer.
It's ok with static libs, but with shared libs all the hard work of maintaining ABI compatibility in libraries is.. wasted?
